# 例外処理

### 例外が起こる実例

```
・数字に対して0で割り算を行う
-> ArithmeticException
・nullのオブジェクトに対してメソッドを呼び出す
-> NullPointerException
・配列の要素数よりも大きい要素数を指定する処理が実行された
-> ArrayIndexOutOfBoundsException
・存在しないファイルを指定し、ファイルの読み込みを行った
-> FileNotFoundException
```

<br>

### try-catch文の基礎構造

```
try {
    // 例外が発生する可能性がある処理
    System.out.println("処理を開始します");
    int result = 10 / 0; // 例: 10を0で割るため、例外が発生
} catch (ArithmeticException e) {
    // 例外が発生した場合の処理
    System.out.println("例外が発生しました: " + e.getMessage());
} finally {
    // 必ず実行される処理
    System.out.println("処理を終了します");
}
```

上のコードを実行すると、以下のように出力されます。

```
処理を開始します
例外が発生しました: / by zero
処理を終了します
```

<br>
<br>


### 各ブロックの説明

```
tryブロック:
例外が発生する可能性のあるコードを記述します。
もしtry内のコードで例外が発生すると、それ以降のtryブロック内のコードは実行されません。
例外が発生しなかった場合、catchブロックの処理は実行されません。

catchブロック:
tryブロックで特定の例外が発生した場合の処理を記述します。
(例外クラス 変数名)の形式で例外をキャッチし、その詳細を利用できます。
catchブロックは複数定義できます

try {
    ...
} catch (例外クラス1 変数名) {
    // 例外クラス1に対する処理
} catch (例外クラス2 変数名) {
    // 例外クラス2に対する処理
}

finallyブロック (任意):
例外の有無にかかわらず、必ず実行されるコードを記述します。
リソースの解放や後処理に使用されることが多いです。
※ 現場で使った記憶はほとんどないです
```

### 検査例外と非検査例外

検査例外とはコンパイル時にチェックされる例外のことです。
メソッド内で発生する可能性がある場合はtry-catch文で処理するか、
throws句を使ってメソッドのシグネチャに例外を記述する必要があります。
Exceptionを継承した例外がこれにあたります。

```java
public void method() throws IOException {
    // 例外が発生する可能性のある処理
    FileInputStream fis = new FileInputStream("file.txt");
}
```

非検査例外とはコンパイル時にチェックされない例外のことです。
プログラムの実行中に発生する可能性がある例外ですが、通常はプログラマのミスによって発生します。
RuntimeExceptionを継承した例外がこれにあたります。

```java
public void method() {
    // 例外が発生する可能性のある処理
    int result = 10 / 0; // ArithmeticException

    String str = null;
    str.length(); // NullPointerException

    int[] arr = new int[5];
    arr[10] = 100; // ArrayIndexOutOfBoundsException
}
```

java api docを見ると、各メソッドには発生する可能性のある例外が記載されています。
- https://docs.oracle.com/javase/jp/17/docs/api/java.base/java/lang/Integer.html#parseInt(java.lang.String)
- https://docs.oracle.com/javase/jp/17/docs/api/java.base/java/nio/file/Files.html#createFile(java.nio.file.Path,java.nio.file.attribute.FileAttribute...)

発生する可能性のある例外をtry-catch文で処理するか、発生しないようにプログラムを修正するかは状況次第です


### 独自例外の実装

検査例外とするか非検査例外とするかは、状況次第です
```java

public class ApplicationException extends Exception {
    public ApplicationException(String message) {
        super(message);
    }
}

public class ApplicationException extends RuntimeException {
    public ApplicationException(String message) {
        super(message);
    }
}
```
